lambda/matching.ml

```diff
@@ -56,409 +58,509 @@ and may_compats = MayCompat.compats
+          | Tpat_any -> (p, rem)
+          | _ -> raise NoMatch
+      )
+    | Tpat_array omegas -> (
+        let len = List.length omegas in
+        fun q rem ->
+          match q.pat_desc with
+          | Tpat_array args when List.length args = len -> (p, args @ rem)
+          | Tpat_any -> (p, omegas @ rem)
+          | _ -> raise NoMatch
+      )
+    | Tpat_tuple omegas -> (
+        let len = List.length omegas in
+        fun q rem ->
+          match q.pat_desc with
+          | Tpat_tuple args when List.length args = len -> (p, args @ rem)
+          | Tpat_any -> (p, omegas @ rem)
+          | _ -> raise NoMatch
+      )
+    | Tpat_record (((_, lbl, _) :: _ as l), _) -> (
+        (* Records are normalized *)
+        let len = Array.length lbl.lbl_all in
+        fun q rem ->
+          match q.pat_desc with
+          | Tpat_record (((_, lbl', _) :: _ as l'), _)
+            when Array.length lbl'.lbl_all = len ->
+              let l' = all_record_args l' in
+              (p, List.fold_right (fun (_, _, p) r -> p :: r) l' rem)
+          | Tpat_any -> (p, List.fold_right (fun (_, _, p) r -> p :: r) l rem)
+          | _ -> raise NoMatch
+      )
+    | Tpat_lazy omega -> (
+        fun q rem ->
+          match q.pat_desc with
+          | Tpat_lazy arg -> (p, arg :: rem)
+          | Tpat_any -> (p, omega :: rem)
+          | _ -> raise NoMatch
+      )
+    | _ -> fatal_error "Matching.Context.matcher"
+
+  let specialize q ctx =
+    let matcher = ctx_matcher q in
+    let rec filter_rec : t -> t = function
+      | ({ right = p :: ps } as l) :: rem -> (
+          match p.pat_desc with
+          | Tpat_or (p1, p2, _) ->
+              filter_rec
+                ({ l with right = p1 :: ps }
+                :: { l with
+                     Row.right (* disam not principal, OK *) = p2 :: ps
+                   }
+                :: rem
+                )
+          | Tpat_alias (p, _, _) ->
+              filter_rec ({ l with right = p :: ps } :: rem)
+          | Tpat_var _ -> filter_rec ({ l with right = omega :: ps } :: rem)
+          | _ -> (
               let rem = filter_rec rem in
               try
-                matcher p ps::rem
-              with
-              | NoMatch -> rem
-              | OrPat   ->
+                let to_left, right = matcher p ps in
+                { left = to_left :: l.left; right } :: rem
+              with NoMatch -> rem
+            )
+        )
+      | [] -> []
+      | _ -> fatal_error "Matching.Context.specialize"
+    in
+    filter_rec ctx
+
+  let select_columns pss ctx =
+    let n = ncols pss in
+    let lub_row ps { Row.left; right } =
+      let transfer, right = rev_split_at n right in
+      match lubs transfer ps with
+      | exception Empty -> None
+      | inter -> Some { Row.left = inter @ left; right }
+    in
+    let lub_with_ctx ps = List.filter_map (lub_row ps) ctx in
+    List.flatten (List.map lub_with_ctx pss)
```

Comment:
The changes are nice but they would be even better with #8760 and its `List.flat_map`.

Reply:
Ah indeed, this should have been `List.concat (List.map ...)`. Pretty sure @gasche is to blame for that one.